src/app/clusters/group-key-mgmt-server/group-key-mgmt-server.cpp

```diff
@@ -0,0 +1,410 @@
+    {
+        fabricIndex = std::get<0>(*it);
+        groupId     = std::get<1>(*it);
+        keySet      = std::get<2>(*it);
+        emberAfGroupKeyManagementClusterPrint("  fabric:%u, group:%u, set:%u", fabricIndex, groupId, keySet);
+    }
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static EmberAfStatus assignKey(uint16_t fabricIndex, GroupId groupId, uint16_t keySetIndex)
+{
+    KeySet * set = getKeySet(fabricIndex, keySetIndex);
+    if (set)
+    {
+        sGroupStates.insert(std::tuple<uint16_t, GroupId, uint16_t>(fabricIndex, groupId, keySetIndex));
+        printStatus();
+        return EMBER_ZCL_STATUS_SUCCESS;
+    }
+    else
+    {
+        return EMBER_ZCL_STATUS_NOT_FOUND;
+    }
+}
+
+static EmberAfStatus revokeKey(uint16_t fabricIndex, GroupId groupId, uint16_t keySetIndex)
+{
+    sGroupStates.erase(std::tuple<uint16_t, GroupId, uint16_t>(fabricIndex, groupId, keySetIndex));
+    printStatus();
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static EmberAfStatus revokeKeys(uint16_t fabricIndex, uint16_t keySetIndex)
+{
+    std::set<std::tuple<uint16_t, GroupId, uint16_t>>::iterator it = sGroupStates.begin();
+    while (it != sGroupStates.end())
+    {
+        auto curr = it++;
+        if (std::get<0>(*curr) == fabricIndex && std::get<2>(*curr) == keySetIndex)
+        {
+            sGroupStates.erase(curr);
+        }
+    }
+    printStatus();
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+static EmberAfStatus revokeKeys(uint16_t fabricIndex)
+{
+    std::set<std::tuple<uint16_t, GroupId, uint16_t>>::iterator it = sGroupStates.begin();
+    while (it != sGroupStates.end())
+    {
+        auto curr = it++;
+        if (std::get<0>(*curr) == fabricIndex)
+        {
+            sGroupStates.erase(curr);
+        }
+    }
+    printStatus();
+    return EMBER_ZCL_STATUS_SUCCESS;
+}
+
+//------------------------------------------------------------------------------
+// Group Key Management Cluster server callbacks
+//------------------------------------------------------------------------------
+
+/** @brief Group Key Management Cluster Server Init
+ *
+ * Server Init
+ *
+ * @param endpoint    Endpoint that is being initialized
+ */
+void emberAfGroupKeyManagementClusterServerInitCallback(chip::EndpointId endpoint) {}
+
+/**
+ * @brief  Cluster AddKey Command callback (from client)
+ */
+bool emberAfGroupKeyManagementClusterSetKeyCallback(chip::EndpointId endpoint, chip::app::CommandHandler * commandObj,
+                                                    uint16_t fabricIndex, uint16_t keySetIndex, chip::ByteSpan keyRoot,
+                                                    chip::ByteSpan epochStartTime, uint8_t securityPolicy)
+{
+    EmberAfStatus result = EMBER_ZCL_STATUS_FAILURE;
+
+    if (0 == endpoint)
+    {
+        result = setKey(fabricIndex, keySetIndex, keyRoot, epochStartTime, securityPolicy);
+    }
+    else
+    {
+        result = EMBER_ZCL_STATUS_INVALID_VALUE;
```

Comment:
I think the correct error here is NotFound (since this cluster can only exist on EndPoint 0). In theory, by construction, this should never happen on any endpoint other than 0, so not sure why this check is here for now. It would be an error to have the RootNode device type node on Endpoint 0, and for GroupKeyManagement cluster associated with it, to also be on a different endpoint.